internal/zinc-classfile/src/main/scala/sbt/internal/inc/classfile/Analyze.scala

```diff
@@ -63,22 +63,34 @@ private[sbt] object Analyze {
     // get class to class dependencies and map back to source to class dependencies
     for ((source, classFiles) <- sourceToClassFiles) {
       analysis.startSource(source)
-      val loadedNonLocalClasses =
-        classFiles.map(c => binaryClassNameToLoadedClass(c.className)).filter(_.getCanonicalName != null)
-      val publicInherited: Map[String, Set[String]] =
-        readAPI(source, loadedNonLocalClasses).groupBy(_._1).mapValues(_.map(_._2))
+      val loadedClasses = classFiles.map(c => binaryClassNameToLoadedClass(c.className))
+      // Local classes are either local, anonymous or inner Java classes
+      val (nonLocalClasses, localClassesOrStale) =
+        loadedClasses.partition(_.getCanonicalName != null)
+
+      // Map local classes to the sources of their enclosing classes
+      val localClassesToSources = localClassesOrStale.flatMap { cls =>
+        val enclosingLoadedCls = Option(cls.getEnclosingClass)
+        val sourceOfEnclosing = enclosingLoadedCls.map(binaryToSourceName(_))
+        sourceOfEnclosing.flatten.map(src => cls.getName -> src).toList
+      }.toMap
+
+      /* Get the mapped source file from a given class name. */
+      def getMappedSource(className: String): Option[String] = {
+        val nonLocalSourceName: Option[String] = for {
+          loadedClass <- binaryClassNameToLoadedClass.get(className)
+          sourceName <- binaryToSourceName(loadedClass)
+        } yield sourceName
+
+        if (nonLocalSourceName.isDefined) nonLocalSourceName
+        else localClassesToSources.get(className)
+      }
 
       def processDependency(onBinaryName: String, context: DependencyContext, fromBinaryName: String): Unit = {
-        val fromClassName = {
-          val clsName = for {
-            loadedClass <- binaryClassNameToLoadedClass.get(fromBinaryName)
-            sourceName <- binaryToSourceName(loadedClass)
-          } yield sourceName
-          clsName match {
-            case Some(resolvedClassName) => resolvedClassName
-            // TODO: handle dependencies coming from local classes the same way as it's done for Scala
-            // classes, see: https://github.com/sbt/sbt/issues/1104#issuecomment-169146039
-            case None                    => return
+        val fromClassName: String = {
+          getMappedSource(fromBinaryName) match {
+            case Some(sourceName) => sourceName
+            case None             => return // It could be a stale class file
```

Comment:
TBH I don't like returns in scala code (but I don't want force my way here :) ). Maybe instead:
```scala
def processDependency(onBinaryName: String, context: DependencyContext, fromBinaryName: String): Unit = 
getMappedSource(fromBinaryName) match {
 case None => 
 case Some(fromClassName) =>
   // Old code here
}
```